arrow-libs/core/arrow-core/src/commonMain/kotlin/arrow/core/raise/Raise.kt

```diff
@@ -0,0 +1,282 @@
+@file:OptIn(ExperimentalTypeInference::class)
+
+package arrow.core.raise
+
+import arrow.core.Either
+import arrow.core.EmptyValue
+import arrow.core.None
+import arrow.core.Option
+import arrow.core.Some
+import arrow.core.Validated
+import arrow.core.continuations.EffectScope
+import arrow.core.identity
+import kotlin.contracts.ExperimentalContracts
+import kotlin.contracts.contract
+import kotlin.experimental.ExperimentalTypeInference
+import kotlin.jvm.JvmName
+
+@DslMarker
+public annotation class EffectDSL
+
+/**
+ * The [Raise] DSL allows you to work with _logical failures_ of type [R].
+ * A _logical failure_ does not necessarily mean that the computation has failed,
+ * but that it has stopped or _short-circuited_.
+ *
+ * The [Raise] DSL allows you to [raise] _logical failure_ of type [R], and you can [recover] from them.
+ *
+ * ```kotlin
+ * fun Raise<String>.failure(): Int = raise("failed")
+ *
+ * fun Raise<Nothing>.recovered(): String =
+ *   recover({ program() }) { failure: String ->
```

Comment:
I guess we want to invoke `failure()` instead of `program()` here
```suggestion
 *   recover({ failure() }) { failure: String ->
```